File: examples/workflows/01-simple-sequential/main.ts
/**
 * Example 1: Simple Sequential Workflow - TypeScript
 *
 * This example shows the most basic AxonFlow workflow: send a query to an LLM and get a response.
 *
 * VALIDATION: This example exits with code 1 if any assertion fails.
 */

import { AxonFlow } from '@axonflow/sdk';

const failures: string[] = [];

function assertCheck(condition: boolean, message: string): void {
  if (condition) {
    console.log(`   PASS: ${message}`);
  } else {
    console.log(`   FAIL: ${message}`);
    failures.push(message);
  }
}

async function main() {
  // Get AxonFlow configuration from environment
  const agentUrl = process.env.AXONFLOW_AGENT_URL || 'http://localhost:8080';
  const clientId = process.env.AXONFLOW_CLIENT_ID;
  const clientSecret = process.env.AXONFLOW_CLIENT_SECRET;

  if (!clientId || !clientSecret) {
    console.error('AXONFLOW_CLIENT_ID and AXONFLOW_CLIENT_SECRET must be set');
    process.exit(1);
  }

  // Create AxonFlow client
  const client = new AxonFlow({
    endpoint: agentUrl,
    clientId,
    clientSecret,
  });

  console.log('Connected to AxonFlow');

  // Define a simple query
  const query = 'What is the capital of France?';
  console.log(`Sending query: ${query}`);

  try {
    // Send query to AxonFlow
    const response = await client.proxyLLMCall({
      userToken: 'user-123',
      query: query,
      requestType: 'chat',
      context: {
        model: 'gpt-4',
      },
    });

    // Print response
    console.log(`Response: ${JSON.stringify(response.data)}`);

    // Assertions - validate actual response structure and content
    console.log('\n--- Assertions ---');
    assertCheck(response !== null && response !== undefined, 'Response object exists');
    assertCheck(response.data !== null && response.data !== undefined, 'Response has data field');
    assertCheck(!response.blocked, 'Response is not blocked');
    assertCheck(
      response.blocked === false || response.blockReason === undefined || response.blockReason === '',
      'No block reason when not blocked'
    );

    // Validate response content contains expected answer
    const responseText = JSON.stringify(response.data).toLowerCase();
    assertCheck(
      responseText.includes('paris') || responseText.includes('france'),
      'Response mentions Paris or France (expected answer)'
    );

    // Validate response metadata
    assertCheck(
      response.requestId !== undefined && response.requestId !== '',
      'Response has a request ID for tracing'
    );

    console.log('\nWorkflow completed successfully');
  } catch (error) {
    console.error(`Query failed: ${error}`);
    process.exit(1);
  }

  process.exit(failures.length > 0 ? 1 : 0);
}

main();

File: examples/workflows/01-simple-sequential/main.py
#!/usr/bin/env python3
"""
Example 1: Simple Sequential Workflow - Python

This example shows the most basic AxonFlow workflow: send a query to an LLM and get a response.
"""

import asyncio
import os
import sys

from axonflow import AxonFlow


async def main():
    # Get AxonFlow configuration from environment
    agent_url = os.getenv("AXONFLOW_ENDPOINT", os.getenv("AXONFLOW_AGENT_URL", "http://localhost:8080"))
    client_id = os.getenv("AXONFLOW_CLIENT_ID", "workflow-example")
    client_secret = os.getenv("AXONFLOW_CLIENT_SECRET", "")

    # Create AxonFlow client
    client = AxonFlow(
        endpoint=agent_url,
        client_id=client_id,
        client_secret=client_secret if client_secret else None,
    )

    print("✅ Connected to AxonFlow")

    # Define a simple query
    query = "What is the capital of France?"
    print(f"📤 Sending query: {query}")

    try:
        # Send query to AxonFlow (async method)
        response = await client.proxy_llm_call(
            user_token=os.getenv("AXONFLOW_USER_TOKEN", "user-123"),
            query=query,
            request_type="chat",
            context={"provider": "openai"},
        )

        # Print response
        if hasattr(response, 'data'):
            print(f"📥 Response: {response.data}")
        elif hasattr(response, 'result'):
            print(f"📥 Response: {response.result}")
        else:
            print(f"📥 Response: {response}")
        print("✅ Workflow completed successfully")
    except Exception as e:
        print(f"❌ Query failed: {e}")
        sys.exit(1)


if __name__ == "__main__":
    asyncio.run(main())
